In [15]:
import pandas as pd
import json
from datetime import datetime

In [3]:
# Loading JSON data into Pandas Dataframes
receipts_path = "C:/Users/naman/OneDrive/Documents/FetchAssignment/data/receipts.json"
brands_path = "C:/Users/naman/OneDrive/Documents/FetchAssignment/data/brands.json"
users_path = "C:/Users/naman/OneDrive/Documents/FetchAssignment/data/users.json"

def read_json_data(file_path):
    with open(file_path, 'r') as f:
        lines = f.readlines()
        data = [json.loads(line) for line in lines]
    return pd.json_normalize(data)

# Reading JSON files
receipts_df = read_json_data(receipts_path)
brands_df = read_json_data(brands_path)
users_df = read_json_data(users_path)

In [19]:
# Date convsersion from Json format to timestamp
# Receipt
receipts_df['createDate.$date'] = pd.to_datetime(receipts_df['createDate.$date'], unit='ms')
receipts_df['dateScanned.$date'] = pd.to_datetime(receipts_df['dateScanned.$date'], unit='ms')
receipts_df['finishedDate.$date'] = pd.to_datetime(receipts_df['finishedDate.$date'], unit='ms')
receipts_df['modifyDate.$date'] = pd.to_datetime(receipts_df['modifyDate.$date'], unit='ms')
receipts_df['pointsAwardedDate.$date'] = pd.to_datetime(receipts_df['pointsAwardedDate.$date'], unit='ms')
receipts_df['purchaseDate.$date'] = pd.to_datetime(receipts_df['purchaseDate.$date'], unit='ms')
# Users
users_df['createdDate.$date'] = pd.to_datetime(users_df['createdDate.$date'], unit='ms')
users_df['lastLogin.$date'] = pd.to_datetime(users_df['lastLogin.$date'], unit='ms')

In [20]:
receipts_df.head(2)

bonusPointsEarned                            bonusPointsEarnedReason  \
0              500.0  Receipt number 2 completed, bonus point schedu...   
1              150.0  Receipt number 5 completed, bonus point schedu...   

  pointsEarned  purchasedItemCount  \
0        500.0                 5.0   
1        150.0                 2.0   

                              rewardsReceiptItemList rewardsReceiptStatus  \
0  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   

  totalSpent                    userId                  _id.$oid  \
0      26.00  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   
1      11.00  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b   

     createDate.$date   dateScanned.$date  finishedDate.$date  \
0 2021-01-03 15:25:31 2021-01-03 15:25:31 2021-01-03 15:25:31   
1 2021-01-03 15:24:43 2021-01-03 15:24:43 2021-01-03 15:24:43   

     modifyDate.$date pointsAwardedDate.$date  purchaseDate.$date  
0 2021-01-03 15:25:36     2021-01-03 15:25:31 2021-01-03 00:00:00  
1 2021-01-03 15:24:48     2021-01-03 15:24:43 2021-01-02 15:24:43

In [28]:
brands_df.head(2)

barcode   category categoryCode                       name topBrand  \
0  511111019862     Baking       BAKING  test brand @1612366101024    False   
1  511111519928  Beverages    BEVERAGES                  Starbucks    False   

                   _id.$oid              cpg.$id.$oid cpg.$ref  brandCode  
0  601ac115be37ce2ead437551  601ac114be37ce2ead437550     Cogs        NaN  
1  601c5460be37ce2ead43755f  5332f5fbe4b03c9a25efd0ba     Cogs  STARBUCKS

In [22]:
users_df.head(2)

active      role signUpSource state                  _id.$oid  \
0    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
1    True  consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   

        createdDate.$date               lastLogin.$date  
0 2021-01-03 15:24:04.800 2021-01-03 15:25:37.857999872  
1 2021-01-03 15:24:04.800 2021-01-03 15:25:37.857999872

In [48]:
print(receipts_df.dtypes,'\n\n', brands_df.dtypes,'\n\n', users_df.dtypes)

bonusPointsEarned                 float64
bonusPointsEarnedReason            object
pointsEarned                       object
purchasedItemCount                float64
rewardsReceiptItemList             object
rewardsReceiptStatus               object
totalSpent                         object
userId                             object
_id.$oid                           object
createDate.$date           datetime64[ns]
dateScanned.$date          datetime64[ns]
finishedDate.$date         datetime64[ns]
modifyDate.$date           datetime64[ns]
pointsAwardedDate.$date    datetime64[ns]
purchaseDate.$date         datetime64[ns]
dtype: object 

 barcode         object
category        object
categoryCode    object
name            object
topBrand        object
_id.$oid        object
cpg.$id.$oid    object
cpg.$ref        object
brandCode       object
dtype: object 

 active                         bool
role                         object
signUpSource                 object
state                

- Value columns like total spend, points earned can be maintained as numerical value rather than string datatype

In [30]:
# Missing values check for critical columns in Receipts
missing_values_summary = {
    'missing_user_id': receipts_df['userId'].isna().sum(),
    'missing_receipt_id': receipts_df['_id.$oid'].isna().sum(),
    'missing_final_price': receipts_df['totalSpent'].isna().sum(),
    'missing_pointsEarned': receipts_df['pointsEarned'].isna().sum()
}

missing_receipts_df = pd.DataFrame([missing_values_summary])
print('Receipts\ntotal_records:', len(receipts_df))
print(missing_receipts_df)

# Missing values check for critical columns in Brands
missing_brands_summary = {
    'missing_barcode': brands_df['barcode'].isna().sum(),
    'missing_brand_id': brands_df['_id.$oid'].isna().sum(),
    'missing_name': brands_df['name'].isna().sum(),
    'missing_brandCode': brands_df['brandCode'].isna().sum()
}

missing_brands_df = pd.DataFrame([missing_brands_summary])
print('Brands\ntotal_records:', len(brands_df))
print(missing_brands_df)

# Missing values check for critical columns in Users
missing_users_summary = {
    'missing_user_id': users_df['_id.$oid'].isna().sum(),
    'missing_state': users_df['state'].isna().sum()
}

missing_users_df = pd.DataFrame([missing_users_summary])
print('Users\ntotal_records:', len(users_df))
print(missing_users_df)

Receipts
total_records: 1119
   missing_user_id  missing_receipt_id  missing_final_price  \
0                0                   0                  435   

   missing_pointsEarned  
0                   510  
Brands
total_records: 1167
   missing_barcode  missing_brand_id  missing_name  missing_brandCode
0                0                 0             0                234
Users
total_records: 495
   missing_user_id  missing_state
0                0             56


- Missing user's money spent for fetch reward transaction as well as the points earned from the recipts scan
- Missing brand code for 234 brands
- Missing user's basic demographics like state

In [36]:
# Checking for duplicate brand IDs
duplicate_brands = brands_df[brands_df.duplicated('_id.$oid', keep=False)]
print(duplicate_brands)

# Checking for duplicate user IDs
duplicate_users = users_df[users_df.duplicated('_id.$oid', keep=False)]
print(duplicate_users)

# Checking for duplicate receipt IDs
duplicate_receipts = receipts_df[receipts_df.duplicated('_id.$oid', keep=False)]
print(duplicate_receipts)

Empty DataFrame
Columns: [barcode, category, categoryCode, name, topBrand, _id.$oid, cpg.$id.$oid, cpg.$ref, brandCode]
Index: []
     active         role signUpSource state                  _id.$oid  \
0      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
1      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
2      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
3      True     consumer        Email    WI  5ff1e1eacfcf6c399c274ae6   
4      True     consumer        Email    WI  5ff1e194b6a9d73a3a9f1052   
..      ...          ...          ...   ...                       ...   
490    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
491    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
492    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
493    True  fetch-staff          NaN   NaN  54943462e4b07e684157a532   
494    True  fetch-staff          NaN   NaN  54943462e4b07e684157a5

- Users data contain duplicates for user ids, on reviewing further all those rows are same, hence, can be deduplicated

In [41]:
# inconsistencies in item list fields

receipts_df.rewardsReceiptItemList.head(7), receipts_df.rewardsReceiptItemList.tail(7)

(0    [{'barcode': '4011', 'description': 'ITEM NOT ...
 1    [{'barcode': '4011', 'description': 'ITEM NOT ...
 2    [{'needsFetchReview': False, 'partnerItemId': ...
 3    [{'barcode': '4011', 'description': 'ITEM NOT ...
 4    [{'barcode': '4011', 'description': 'ITEM NOT ...
 5    [{'barcode': '4011', 'description': 'ITEM NOT ...
 6    [{'brandCode': 'MISSION', 'competitorRewardsGr...
 Name: rewardsReceiptItemList, dtype: object,
 1112    [{'barcode': 'B076FJ92M4', 'description': 'mue...
 1113    [{'barcode': 'B076FJ92M4', 'description': 'mue...
 1114    [{'barcode': 'B076FJ92M4', 'description': 'mue...
 1115                                                  NaN
 1116                                                  NaN
 1117    [{'barcode': 'B076FJ92M4', 'description': 'mue...
 1118                                                  NaN
 Name: rewardsReceiptItemList, dtype: object)

- Inconsistencies and missing infromation in item details in receipts. This affects the Receipt Item data and it's connection to Brands data. The column fields are not consistent for each receipt, hence require default values for the missing content in the database.

In [49]:
receipts_df.head(1)

bonusPointsEarned                            bonusPointsEarnedReason  \
0              500.0  Receipt number 2 completed, bonus point schedu...   

  pointsEarned  purchasedItemCount  \
0        500.0                 5.0   

                              rewardsReceiptItemList rewardsReceiptStatus  \
0  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   

  totalSpent                    userId                  _id.$oid  \
0      26.00  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   

     createDate.$date   dateScanned.$date  finishedDate.$date  \
0 2021-01-03 15:25:31 2021-01-03 15:25:31 2021-01-03 15:25:31   

     modifyDate.$date pointsAwardedDate.$date purchaseDate.$date  
0 2021-01-03 15:25:36     2021-01-03 15:25:31         2021-01-03

In [54]:
# data ranges validity
print(receipts_df['createDate.$date'].min(), receipts_df['createDate.$date'].max())
print(receipts_df['dateScanned.$date'].min(), receipts_df['dateScanned.$date'].max())
print(receipts_df['finishedDate.$date'].min(), receipts_df['finishedDate.$date'].max())
print(receipts_df['modifyDate.$date'].min(), receipts_df['modifyDate.$date'].max())
print(receipts_df['pointsAwardedDate.$date'].min(), receipts_df['pointsAwardedDate.$date'].max())
print(receipts_df['purchaseDate.$date'].min(), receipts_df['purchaseDate.$date'].max())

2020-10-30 20:17:59 2021-03-01 23:17:34.772000
2020-10-30 20:17:59 2021-03-01 23:17:34.772000
2021-01-03 15:24:10 2021-02-26 22:36:25
2021-01-03 15:24:10 2021-03-01 23:17:34.772000
2020-10-30 20:18:00 2021-02-26 22:36:25
2017-10-30 00:00:00 2021-03-08 17:37:13


- Date ranges look valid except the record having a purchase date of year 2017. Let's check further.

In [61]:
receipts_df[receipts_df['purchaseDate.$date']=='2017-10-30 00:00:00'].head()

bonusPointsEarned                            bonusPointsEarnedReason  \
0              500.0  Receipt number 2 completed, bonus point schedu...   
1              150.0  Receipt number 5 completed, bonus point schedu...   
2                5.0                         All-receipts receipt bonus   
3                5.0                         All-receipts receipt bonus   
4                5.0                         All-receipts receipt bonus   

  pointsEarned  purchasedItemCount  \
0        500.0                 5.0   
1        150.0                 2.0   
2            5                 1.0   
3          5.0                 4.0   
4          5.0                 2.0   

                              rewardsReceiptItemList rewardsReceiptStatus  \
0  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2  [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   

  totalSpent                    userId                  _id.$oid  \
0      26.00  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575   
1      11.00  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b   
2      10.00  5ff1e1f1cfcf6c399c274b0b  5ff1e1f10a720f052300057a   
3      28.00  5ff1e1eacfcf6c399c274ae6  5ff1e1ee0a7214ada100056f   
4       1.00  5ff1e194b6a9d73a3a9f1052  5ff1e1d20a7214ada1000561   

     createDate.$date   dateScanned.$date  finishedDate.$date  \
0 2021-01-03 15:25:31 2021-01-03 15:25:31 2021-01-03 15:25:31   
1 2021-01-03 15:24:43 2021-01-03 15:24:43 2021-01-03 15:24:43   
2 2021-01-03 15:25:37 2021-01-03 15:25:37                 NaT   
3 2021-01-03 15:25:34 2021-01-03 15:25:34 2021-01-03 15:25:34   
4 2021-01-03 15:25:06 2021-01-03 15:25:06 2021-01-03 15:25:11   

     modifyDate.$date pointsAwardedDate.$date   purchaseDate.$date  \
0 2021-01-03 15:25:36     2021-01-03 15:25:31  2017-10-30 00:00:00   
1 2021-01-03 15:24:48     2021-01-03 15:24:43  2017-10-30 00:00:00   
2 2021-01-03 15:25:42                     NaT  2017-10-30 00:00:00   
3 2021-01-03 15:25:39     2021-01-03 15:25:34  2017-10-30 00:00:00   
4 2021-01-03 15:25:11     2021-01-03 15:25:06  2017-10-30 00:00:00   

   2017-10-30 00:00:00  
0  2017-10-30 00:00:00  
1  2017-10-30 00:00:00  
2  2017-10-30 00:00:00  
3  2017-10-30 00:00:00  
4  2017-10-30 00:00:00

It looks like a valid data, howver, it seems odd business-wise as users earned rewards for submitting ~4 years old reciepts